ADR/ADR-003-PyBind11-For-CPP-Bindings.md

```diff
@@ -48,248 +54,364 @@ However, this approach presents several technical and business challenges:
- * Cons: higher compile-time due to code generation, larger binary size
+Python use has become ever more widespread thus making functionality of our C++
+stack available has become an important consideration. We are looking for a
+solution that:
+- Offers us to provide Python bindings that integrate well into the Python
+  environment.
+- Are stable, mature and have good long term maintenance outlook.
+- Add the least amount of overall complexity for the most amount of development
+  ease.
 
-Nanobind:
- * Pros: compensates pybind11 cons
- * Cons: not header-only anymore, fewer `std` or third party support
+Overall PyBind11 fits best to the outlined criteria. PyBind11 is broadly
+adopted, has a large contributor community and is one of the most active
+developed options considered.
 
-Boost.Python:
- * Cons: clunky, heavy dependency on boost, larger build complexity, slower compile times
+PyBind11 adds delivery complexity by requiring distribution of binary wheels
+per Python minor version. Of the considered alternatives only Nanobind (Stable
+API support) CFFI, and manually writing bindings could provide universal binary
+wheels. This added complexity is offset by the plethora of developer
+convenience features PyBind11 offers. Most noteworthy here is that is comes
+with a lot of pre-defined conversions for types from std. PyBind11 offers the
+widest support in this category.
 
-Swig:
- * Pros: automates most of the binding work, supports multiple languages, mature and stable tool
- * Cons: interface files add an extra step, generated code may be harder to debug, 
-   less "pythonic" feel compared to pybind11 or Boost.Python
 
+CPython and SWIG are discarded for their low level abstractions and large
+amount of rote and error-prone development resulting from this.
 
-Our canditate of choice is Pybind11 - compared to all the other candidates (except nanobind) it offers the 
-  * lowest development complexity
-  * gives direct access to complex objects like iterators
-  * is well maintained & documented
-  * battle-tested and widely used in famous frameworks
+CFFI is discarded because of the additional C-API that is required to build and
+the additional python code that turns Procedural APIs back into Object Oriented
+APIs.
 
-Depending on the development on nanobind over the next coming years. 
-It might become a replacement for pybind11 - as the API is similar to pybind11, a migration will be rather simple.
+Boost.Python is primarily discarded because we do not want to pull in Boost.
+Secondarily because it less fully featured than PyBind11.
 
+Nanobind is less fully featured and explicitly chooses to not provide features
+we want to use, i.e. module local bindings. Additionally it is a comparatively
+young project with a smaller contributor base.
+
+Cython is similar to SWIG and CFFI a more verbose and manual process, although
+to a less extend. Further it does not provide automatic type conversions as
```

Comment:
Not sure I agree with that part. There's not that big a difference between writing Cython code that wraps a C++ API and writing C++ code that uses PyBind11 to expose the API. And I would even dare to say that I find Cython code more readable than C++ code using PyBind11, and while this is partly because I am doing much more Python than C++ these days, I will not be the only one in that situation.